crates/sui-framework/sources/safe.move

```diff
@@ -0,0 +1,214 @@
+    }
+
+    //////////////////////////////////////////////////////
+    /// PUBLIC FUNCTIONS
+    //////////////////////////////////////////////////////
+
+    /// Wrap a coin around a safe.
+    /// a trusted party (or smart contract) to transfer the object out.
+    public fun create<T>(balance: Balance<T>, ctx: &mut TxContext) {
+        let box = Safe {
+            id: object::new(ctx),
+            balance,
+            owner: sender(ctx),
+            next_nonce: 0,
+            permissions_bitmap: vector::empty(),
+        };
+        transfer::share_object(box)
+    }
+
+    /// Wraps an object with type T into a lootbox, and returns a transfer capability. Useful for creating a safe and
+    /// providing permissions to an external party/smart contract within one transaction.
+    public fun create_with_capability<T>(balance: Balance<T>, capability_amount: u64, ctx: &mut TxContext): Capability<T> {
+        let safe = Safe {
+            id: object::new(ctx),
+            balance,
+            owner: sender(ctx),
+            next_nonce: 0,
+            permissions_bitmap: vector::empty(),
+        };
+
+        let capability = Capability {
+            id: object::new(ctx),
+            safe_id: object::uid_to_inner(&safe.id),
+            amount: capability_amount,
+            nonce: 0,
+        };
+
+        transfer::share_object(safe);
+        capability
+    }
+
+    /// Withdraw coins from the safe as an owner.
+    public fun withdraw_as_owner<T>(safe: &mut Safe<T>, withdraw_amount: u64, ctx: &mut TxContext): Balance<T> {
+        // Ensures that only the owner can withdraw from the safe.
+        assert!(sender(ctx) == safe.owner, 0);
+        balance::split(&mut safe.balance, withdraw_amount)
+    }
+
+    /// CAPABILITY OWNER ONLY
+    /// Withdraw coins from the safe as a capability holder.
+    public fun withdraw_as_capability_holder<T>(safe: &mut Safe<T>, withdraw_amount: u64, capability: &mut Capability<T>): Balance<T> {
+        // Check that the capability pointer matches the UID of the lootbox. Abort if that is not the case.
+        assert!(object::uid_to_inner(&safe.id) == capability.safe_id, 0);
+
+        // Update the funds. Note that this will revert if there aren't sufficient funds.
+        capability.amount = capability.amount - withdraw_amount;
+
+        balance::split(&mut safe.balance, withdraw_amount)
+    }
+
+    /// Safe Owner Only
+    public fun revoke_capability<T>(safe: &mut Safe<T>, capability: &Capability<T>, ctx: &TxContext) {
+        // Ensures that only the owner can withdraw from the safe.
+        assert!(sender(ctx) == safe.owner, 0);
+        bitmap_set(&mut safe.permissions_bitmap, capability.nonce, false);
+    }
+
+    /// Safe Owner Only
+    public fun create_capability<T>(safe: &mut Safe<T>, withdraw_amount: u64, ctx: &mut TxContext): Capability<T> {
+        assert!(sender(ctx) == safe.owner, 0);
+        create_capability_(safe, withdraw_amount, ctx)
+    }
+}
+
```

Comment:
Might also be useful to have API's for:
- allowing the owner (or more likely, anyone?) to deposit more funds into a `Safe`
- allowing a `Capability` owner to voluntarily relinquish permissions
- allowing anyone to read a `Safe`'s `balance`
- deleting an empty `Safe`

Reply:
1 is just pushed in through another commit. The rest I'll add!